### ISSUE TITLE
Autoimports: sqlite3.OperationalError: database is locked

### ISSUE THREAD

**Issue by outp1** [python-lsp/python-lsp-server#529]
Hi. My autoimports didn't work well, so i tried to inspect my problem. All i found is error logs from lsp server like this:

```
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	'Exception in thread Thread-1 (_reload_cache):\nTraceback (most recent call last):\n  File "/home/danya/.pyenv/versions/3.12.1/lib/python3.12/threading.py", line 1073, in _bootstrap_inner\n'
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	"    self.run()\n"
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	'  File "/home/danya/.pyenv/versions/3.12.1/lib/python3.12/threading.py", line 1010, in run\n'
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	"    self._target(*self._args, **self._kwargs)\n"
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	'  File "/home/danya/.local/share/nvim/mason/packages/python-lsp-server/venv/lib/python3.12/site-packages/pylsp/plugins/rope_autoimport.py", line 71, in _reload_cache\n'
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	'    autoimport.generate_cache(task_handle=task_handle, resources=resources)\n  File "/home/danya/.local/share/nvim/mason/packages/python-lsp-server/venv/lib/python3.12/site-packages/rope/contrib/autoimport/sqlite.py", line 368, in generate_cache\n'
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	'    self._execute(\n  File "/home/danya/.local/share/nvim/mason/packages/python-lsp-server/venv/lib/python3.12/site-packages/rope/contrib/autoimport/sqlite.py", line 643, in _execute\n'
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	"    return self.connection.execute(query._query, *args, **kwargs)\n"
[ERROR][2024-02-14 11:55:21] .../vim/lsp/rpc.lua:734	"rpc"	"/home/danya/.local/share/nvim/mason/bin/pylsp"	"stderr"	"           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\nsqlite3.OperationalError: database is locked\n"
```

No related problems was found, except of recently works on something similar like #474 and #499

Any thoughts? I'm not a pylsp developer, so help would be really appreciated.

**Comment by rodda-kyusu** [comment]
I can replicate this problem.

Looks like we are waiting on [this PR](https://github.com/python-lsp/python-lsp-server/pull/530), based on discussion [here](https://github.com/python-lsp/python-lsp-server/issues/374) that references this thread.

Thanks all!

```[stderr] Exception in thread Thread-38 (_reload_cache):
[stderr] Traceback (most recent call last):
[stderr]   File "/Users/rjohn/.pyenv/versions/3.11.6/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
[server-notification] Mon Mar 11 09:06:15 2024:
(:jsonrpc "2.0" :method "$/progress" :params
	  (:token "9ccbdcf4-7c06-4c6d-b01b-ff11ca765147" :value
		  (:kind "end")))
[stderr]     self.run()
[stderr]   File "/Users/rjohn/.pyenv/versions/3.11.6/lib/python3.11/threading.py", line 982, in run
[stderr]     self._target(*self._args, **self._kwargs)
[stderr]   File "/Users/rjohn/.pyenv/versions/3.11.6/envs/core/lib/python3.11/site-packages/pylsp/plugins/rope_autoimport.py", line 71, in _reload_cache
[stderr]     autoimport.generate_cache(task_handle=task_handle, resources=resources)
[stderr]   File "/Users/rjohn/.pyenv/versions/3.11.6/envs/core/lib/python3.11/site-packages/rope/contrib/autoimport/sqlite.py", line 368, in generate_cache
[stderr]     self._execute(
[stderr]   File "/Users/rjohn/.pyenv/versions/3.11.6/envs/core/lib/python3.11/site-packages/rope/contrib/autoimport/sqlite.py", line 643, in _execute
[stderr]     return self.connection.execute(query._query, *args, **kwargs)
[stderr]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[stderr] sqlite3.OperationalError: database is locked```

**Comment by outp1** [comment]
I'm not sure if the problem was actually resolved a couple of versions ago or not, but I now encounter the same problem:

```
[ERROR][2024-12-13 11:22:09] .../vim/lsp/rpc.lua:770	"rpc"	"/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/bin/pylsp"	"stderr"	'2024-12-13 11:22:09,752 MSK - WARNING - pylsp.config.config - Failed to load hook pylsp_document_did_save: database is locked\nTraceback (most recent call last):\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pylsp/config/config.py", line 39, in _hookexec\n    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pluggy/_manager.py", line 480, in traced_hookexec\n    return outcome.get_result()\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pluggy/_result.py", line 100, in get_result\n    raise exc.with_traceback(exc.__traceback__)\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pluggy/_result.py", line 62, in from_call\n    result = func()\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pluggy/_manager.py", line 477, in <lambda>\n    lambda: oldcall(hook_name, hook_impls, caller_kwargs, firstresult)\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pluggy/_callers.py", line 139, in _multicall\n    raise exception.with_traceback(exception.__traceback__)\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pluggy/_callers.py", line 103, in _multicall\n    res = hook_impl.function(*args)\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pylsp/plugins/rope_autoimport.py", line 390, in pylsp_document_did_save\n    cache.reload_cache(config, workspace, [document])\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pylsp/plugins/rope_autoimport.py", line 55, in reload_cache\n    self._reload_cache(workspace, autoimport, resources)\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/pylsp/plugins/rope_autoimport.py", line 71, in _reload_cache\n    autoimport.generate_cache(task_handle=task_handle, resources=resources)\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/rope/contrib/autoimport/sqlite.py", line 390, in generate_cache\n    self._execute(\n  File "/home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/lib/python3.10/site-packages/rope/contrib/autoimport/sqlite.py", line 669, in _execute\n    return self.connection.execute(query._query, *args, **kwargs)\nsqlite3.OperationalError: database is locked\n'
```

To make sure that the version is correct:
```
~ ➤ /home/danya/.local/share/virtualenvs/eco-core-kip-Ca17/bin/pylsp --version
pylsp v1.12.0
```

**Comment by outp1** [comment]
@ccordoba12  Can you reopen it?

**Comment by ccordoba12** [comment]
This is referenced in our Changelog, so we can't reopen it. Please open a new issue about it.
